clients/src/main/java/org/apache/kafka/clients/producer/internals/ProducerBatch.java

```diff
@@ -281,15 +280,14 @@ else if (this.inRetry() && requestTimeoutMs < (waitedTimeMs(now) - retryBackoffM
     }
 
     /**
-     * Completes the produce future with timeout exception and invokes callbacks.
-     * This method should be invoked only if {@link #maybeExpire(int, long, long, long, boolean)}
-     * returned true.
+     * If {@link #maybeExpire(int, long, long, long, boolean)} returned true, the sender will fail the batch with
+     * the exception returned by this method.
+     * @return An exception indicating the batch expired.
      */
-    void expirationDone() {
+    RuntimeException timeoutException() {
```

Comment:
We may as well use the more specific type `TimeoutException` given the name of the method.